x/staking/src/types/domain.rs

```diff
@@ -323,11 +324,9 @@ impl Validator {
         } else {
             // leave excess tokens in the validator
             // however fully use all the delegator shares
-            // TODO: infallible + floor
             let tokens = self.tokens_from_shares(del_shares).unwrap().to_uint_floor();
-            // TODO: check of negative result
+            // the library panics on substruct with overflow and this behavior is identical to sdk
```

Comment:
Same comment about panic

Reply:
I guess this should never happen unless there's a bug somewhere? In which case `expect` it. My view is, if there's something that doesn't add up, better to halt the chain.

Reply:
I don't think that halting node due to an issue with (for example) a transaction for one account is too much. Better return error for this tx

Reply:
I depends... if the problem is caused by the chain being in an "impossible/unreachable" state, even if it was found to be in such a state during tx processing, then the chain should be halted.

Reply:
I checked the method. It is called from fallible method. But the case is rather `impossible/unreachable` because a use can't have a negative balance.